src/DataLevelService.php

```diff
@@ -0,0 +1,312 @@
+      '#default_value' => $settings ? $settings->status : '0',
+    ];
+    $form['bluesnap_data_level_settings']['type'] = [
+      '#type' => 'radios',
+      '#title' => $this->t('Bluesnap data processing level'),
+      '#options' => [
+        'level_2' => $this->t('Level 2'),
+        'level_3' => $this->t('Level 3'),
+      ],
+      '#states' => [
+        'visible' => [
+          ':input[name="bluesnap_data_level_settings[bluensnap_data_level_status]"]' => [
+            'checked' => TRUE,
+          ],
+        ],
+      ],
+      '#default_value' => $settings ? $settings->type : 'level_2',
+    ];
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDataLevelSetting(StoreInterface $store) {
+    $settings = $store->get('bluesnap_data_level_settings')->value;
+    $settings = json_decode($settings);
+    return $settings;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCardSuppprtLevel3() {
+    return ['mastercard', 'visa'];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCardSuppprtLevel2() {
+    return ['mastercard', 'visa', 'amex'];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getData(OrderInterface $order, $card_type) {
+    $store = $order->getStore();
+    $output = [];
+    // Get data level settings for store.
+    $settings = $this->getDataLevelSetting($store);
+    if (!($settings->status)) {
+      return [];
+    }
+
+    // Check the data level setting and
+    // return level2 or level3 data.
+    $data_type = $settings->type;
+    if ($data_type == 'level_3' && in_array($card_type, $this->getCardSuppprtLevel3())) {
+      $output['level3Data'] = $this->level3Data($order, $card_type);
+      return $output;
+    }
+    if ($data_type == 'level_2' && in_array($card_type, $this->getCardSuppprtLevel2())) {
+      $output['level2Data'] = $this->level2Data($order, $card_type);
+      return $output;
+    }
+    return $output;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  private function level2Data(OrderInterface $order, $card_type) {
+    $level_2_data = [];
+    $level_2_data['customerReferenceNumber'] = $this->getCustomerReferenceNumber($order);
+
+    if ($sales_tax = $this->getOrderAdjustment($order, 'tax')) {
+      $level_2_data['salesTaxAmount'] = $sales_tax;
+    }
+
+    if ($card_type != 'amex') {
+      return $level_2_data;
+    }
+
+    // Amex Level 2 with TAA (Transaction Advice Addendum) requires.
+    // destinationZipCode along with level2Data. This is an Amex-specific
+    // level that contains item data (such as item description and quantity)
+    // in addition to Level 2 fields. Only lineItemTotal, description,
+    // itemQuantity are required.
+    if ($shipping_info = $this->getShippingInfo($order)) {
+      $level_2_data['destinationZipCode'] = $shipping_info['destinationZipCode'];
+    }
+    $level_2_data['level3DataItem'] = $this->level3DataItems($order, $card_type);
```

Comment:
Why do we need to pass the level3DataItem in the level 2 data?

Reply:
For Amex card it requires level 3 data items in level 2. I have added a comment above that code now